clients/src/main/java/org/apache/kafka/clients/FetchSessionHandler.java

```diff
@@ -0,0 +1,340 @@
+         * than a regular Map.
+         *
+         * One reason is that when dealing with FULL fetch requests, if there is not enough response
+         * space to return data from all partitions, the server will only return data from partitions
+         * early in this list.
+         *
+         * Another reason is because we make use of the list ordering to optimize the preparation of
+         * incremental fetch requests (see below).
+         */
+        private LinkedHashMap<TopicPartition, PartitionData> next = new LinkedHashMap<>();
+
+        /**
+         * Mark that we want to remove a partition from the incremental fetch session.
+         */
+        public void remove(TopicPartition topicPartition) {
+            next.remove(topicPartition);
+        }
+
+        /**
+         * Mark the data which we want from this partition on the upcoming fetch.
+         */
+        public void add(TopicPartition topicPartition, PartitionData data) {
+            next.put(topicPartition, data);
+        }
+
+        public FetchRequestData build() {
+            if (nextMetadata.isFull()) {
+                log.debug("Built full fetch {} for node {} with {} partition(s).",
+                    nextMetadata, node, next.size());
+                sessionParts = next;
+                next = null;
+                Map<TopicPartition, PartitionData> toSend =
+                    Collections.unmodifiableMap(new LinkedHashMap<>(sessionParts));
+                return new FetchRequestData(toSend, Collections.<TopicPartition>emptyList(), toSend, nextMetadata);
+            }
+            int numAdded = 0, numRemoved = 0, numAltered = 0;
+            List<TopicPartition> toForget = new ArrayList<>(0);
+            for (Iterator<Entry<TopicPartition, PartitionData>> iter =
+                     sessionParts.entrySet().iterator(); iter.hasNext(); ) {
+                Entry<TopicPartition, PartitionData> entry = iter.next();
+                TopicPartition topicPartition = entry.getKey();
+                PartitionData prevData = entry.getValue();
+                PartitionData nextData = next.get(topicPartition);
+                if (nextData != null) {
+                    if (prevData.equals(nextData)) {
+                        // Omit this partition from the FetchRequest, because it hasn't changed
+                        // since the previous request.
+                        next.remove(topicPartition);
+                    } else {
+                        // Move the altered partition to the end of 'next'
+                        next.remove(topicPartition);
+                        next.put(topicPartition, nextData);
+                        entry.setValue(nextData);
+                        numAltered++;
+                    }
+                } else {
+                    // Set maxBytes to 0 to indicate that we no longer want to listen to this partition.
```

Comment:
This comment is no longer relevant, right?

Reply:
Fixed